src/value.c

```diff
@@ -76,26 +76,82 @@ SIValue SI_TransferStringVal(char *s) {
 	};
 }
 
-SIValue SI_Clone(SIValue v) {
-	if(v.type == T_STRING) {
-		// Allocate a new copy of the input's string value
+/* Make an SIValue that reuses the original's allocations, if any.
+ * The returned value is not responsible for freeing any allocations,
+ * and is not guaranteed that these allocations will remain in scope. */
+SIValue SI_ShareValue(const SIValue v) {
+	SIValue dup = v;
+	// If the original value owns an allocation, mark that the duplicate shares it.
+	if(v.allocation == M_SELF) dup.allocation = M_VOLATILE;
+	return dup;
+}
+
+
+/* Make an SIValue that creates its own copies of the original's allocations, if any.
+ * This is not a deep clone: if the inner value holds its own references,
+ * such as the Entity pointer to the properties of a Node or Edge, those are unmodified. */
+SIValue SI_CloneValue(const SIValue v) {
+	if(v.allocation == M_NONE) return v; // Stack value; no allocation necessary.
+
+	if(v.type & SI_STRING) {
+		// Allocate a new copy of the input's string value.
 		return SI_DuplicateStringVal(v.stringval);
 	}
-	SIValue dup;
-	memcpy(&dup, &v, sizeof(SIValue));
-	return dup;
+
+	// Copy the memory region for Node and Edge values. This does not modify the
+	// inner Entity pointer to the value's properties.
+	// TODO None of this logic should be accessible right now
+	SIValue clone;
+	clone.type = v.type;
+	clone.allocation = M_SELF;
+
+	size_t size;
+	if(v.type == T_NODE) {
+		size = sizeof(Node);
+	} else if(v.type == T_EDGE) {
+		size = sizeof(Edge);
+	} else {
+		assert(false && "Encountered heap-allocated SIValue of unhandled type");
+	}
+	clone.ptrval = rm_malloc(size);
+	memcpy(clone.ptrval, v.ptrval, size);
+	return clone;
 }
 
-SIValue SI_ShallowCopy(SIValue v) {
-	SIValue dup = v;
-	// If the original value owns an allocation, mark that the duplicate shares it
-	if(v.allocation == M_SELF) dup.allocation = M_CONST;
-	return dup;
+/* Ensure that any allocation held by the given SIValue is guaranteed to not go out
+ * of scope during the lifetime of this query by copying references to volatile memory.
+ * Heap allocations that are not scoped to the input SIValue, such as strings from the AST
+ * or a GraphEntity property, are not modified. */
+void SIValue_Persist(SIValue *v) {
+	// Do nothing for non-volatile values.
+	if(v->allocation != M_VOLATILE) return;
+
+	// Copy volatile string values.
```

Comment:
Seems like we can reduce all of the following logic with
*v = SI_CloneValue(*v);

Reply:
:+1: Good call!